components/engine_traits/src/flush.rs

```diff
@@ -72,25 +74,51 @@ impl Default for SstApplyState {
     }
 }
 
+#[derive(Debug)]
+pub struct SstApplyEntry {
+    pub applied_index: u64,
+    pub sst: SstMeta,
+}
+
+impl SstApplyEntry {
+    pub fn new(applied_index: u64, sst: SstMeta) -> Self {
+        Self { applied_index, sst }
+    }
+}
+
 impl SstApplyState {
     #[inline]
-    pub fn registe_ssts(&self, uuids: Vec<Vec<u8>>, sst_applied_index: u64) {
+    pub fn register_ssts(&self, applied_index: u64, ssts: Vec<SstMeta>) {
         let mut map = self.sst_map.write().unwrap();
-        for uuid in uuids {
-            map.insert(uuid, sst_applied_index);
+        for sst in ssts {
+            let cf_index = data_cf_offset(sst.get_cf_name());
+            let entry = SstApplyEntry::new(applied_index, sst);
+            map.entry(cf_index).or_default().push(entry);
         }
     }
 
-    /// Query the sst applied index.
     #[inline]
-    pub fn sst_applied_index(&self, uuid: &Vec<u8>) -> Option<u64> {
-        self.sst_map.read().unwrap().get(uuid).copied()
+    pub fn stale_ssts(&self, cf: &str, flushed_index: u64) -> Vec<SstMeta> {
+        let map = self.sst_map.read().unwrap();
+        let cf_index = data_cf_offset(cf);
+        if let Some(ssts) = map.get(&cf_index) {
+            return ssts
+                .iter()
+                .filter(|entry| entry.applied_index <= flushed_index)
+                .map(|entry| entry.sst.clone())
+                .collect();
+        }
+        vec![]
     }
 
-    pub fn delete_ssts(&self, uuids: Vec<Vec<u8>>) {
-        let mut map = self.sst_map.write().unwrap();
-        for uuid in uuids {
-            map.remove(&uuid);
+    pub fn delete_ssts(&self, ssts: &Vec<SstMeta>) {
+        let mut map: std::sync::RwLockWriteGuard<'_, HashMap<usize, Vec<SstApplyEntry>>> =
+            self.sst_map.write().unwrap();
```

Comment:
Can we skip type hint?

```suggestion
        let mut map = self.sst_map.write().unwrap();
```